### ISSUE TITLE
__PACKAGE__ does not work as a hash key

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13074]
Migrated from [rt.perl.org#118701](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=118701) (status was 'rejected')

Searchable as RT118701$

**Comment by p5pRT** [comment]
### From @jimav


This is a bug report for perl from james\_avera@&#8203;yahoo\.com\,
generated with the help of perlbug 1\.39 running under perl 5\.14\.2\.




---
\#\!/usr/bin/perl
use strict; use warnings;

sub show\($\) \{ defined\($\_\[0\]\) ? \("'"\.$\_\[0\]\."'"\) : "undef" \}

my %table = \( main => "the\-value" \);

warn "\_\_PACKAGE\_\_ = "\,show\(\_\_PACKAGE\_\_\)\,"\\n";
warn "table\{\_\_PACKAGE\_\_\} = "\, show\($table\{\_\_PACKAGE\_\_\}\)\, "\\n";
warn "table\{'main'\} = "\, show\($table\{'main'\}\)\, "\\n";

die "bug" unless show\($table\{\_\_PACKAGE\_\_\}\) eq show\($table\{main\}\);
\# This dies



---


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=low


---
Site configuration information for perl 5\.14\.2&#8203;:

Configured by Debian Project at Mon Apr 15 15&#8203;:34&#8203;:46 UTC 2013\.

Summary of my perl5 \(revision 5 version 14 subversion 2\) configuration&#8203;:
&nbsp;   
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=3\.2\.0\-37\-generic\, archname=x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    uname='linux lamiak 3\.2\.0\-37\-generic \#58\-ubuntu smp thu jan 24 15&#8203;:28&#8203;:10 utc 2013 x86\_64 x86\_64 x86\_64 gnulinux '
&nbsp;    config\_args='\-Dusethreads \-Duselargefiles \-Dccflags=\-DDEBIAN \-D\_FORTIFY\_SOURCE=2 \-g \-O2 \-fstack\-protector \-\-param=ssp\-buffer\-size=4 \-Wformat \-Werror=format\-security \-Dldflags= \-Wl\,\-Bsymbolic\-functions \-Wl\,\-z\,relro \-Dlddlflags=\-shared \-Wl\,\-Bsymbolic\-functions \-Wl\,\-z\,relro \-Dcccdlflags=\-fPIC \-Darchname=x86\_64\-linux\-gnu \-Dprefix=/usr \-Dprivlib=/usr/share/perl/5\.14 \-Darchlib=/usr/lib/perl/5\.14 \-Dvendorprefix=/usr \-Dvendorlib=/usr/share/perl5 \-Dvendorarch=/usr/lib/perl5 \-Dsiteprefix=/usr/local \-Dsitelib=/usr/local/share/perl/5\.14\.2 \-Dsitearch=/usr/local/lib/perl/5\.14\.2 \-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3 \-Dsiteman1dir=/usr/local/man/man1 \-Dsiteman3dir=/usr/local/man/man3 \-Duse64bitint \-Dman1ext=1 \-Dman3ext=3perl \-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Ud\_ualarm \-Uusesfio \-Uusenm \-Ui\_libutil \-DDEBUGGING=\-g \-Doptimize=\-O2 \-Duseshrplib \-Dlibperl=libperl\.so\.5\.14\.2 \-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fstack\-protector \-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fstack\-protector \-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.7\.3'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib /usr/lib
&nbsp;    libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    libc=\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.14\.2
&nbsp;    gnulibc\_version='2\.17'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-L/usr/local/lib \-fstack\-protector'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl 5\.14\.2&#8203;:
&nbsp;    /home/t/lib/perl
&nbsp;    /etc/perl
&nbsp;    /usr/local/lib/perl/5\.14\.2
&nbsp;    /usr/local/share/perl/5\.14\.2
&nbsp;    /usr/lib/perl5
&nbsp;    /usr/share/perl5
&nbsp;    /usr/lib/perl/5\.14
&nbsp;    /usr/share/perl/5\.14
&nbsp;    /usr/local/lib/site\_perl
&nbsp;    \.



---
Environment for perl 5\.14\.2&#8203;:
&nbsp;    HOME=/home/t
&nbsp;    LANG=en\_US\.UTF\-8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/t/bin&#8203;:/usr/lib/lightdm/lightdm&#8203;:/usr/local/sbin&#8203;:/usr/local/bin&#8203;:/usr/sbin&#8203;:/usr/bin&#8203;:/sbin&#8203;:/bin&#8203;:/usr/games&#8203;:/usr/local/games
&nbsp;    PERL5LIB=/home/t/lib/perl&#8203;:
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Jun 29 23&#8203;:53&#8203;:13 2013\, jimav wrote&#8203;:
> 
> This is a bug report for perl from james\_avera@&#8203;yahoo\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.14\.2\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \#\!/usr/bin/perl
> use strict; use warnings;
> 
> sub show\($\) \{ defined\($\_\[0\]\) ? \("'"\.$\_\[0\]\."'"\) : "undef" \}
> 
> my %table = \( main => "the\-value" \);
> 
> warn "\_\_PACKAGE\_\_ = "\,show\(\_\_PACKAGE\_\_\)\,"\\n";
> warn "table\{\_\_PACKAGE\_\_\} = "\, show\($table\{\_\_PACKAGE\_\_\}\)\, "\\n";
> warn "table\{'main'\} = "\, show\($table\{'main'\}\)\, "\\n";
> 
> die "bug" unless show\($table\{\_\_PACKAGE\_\_\}\) eq show\($table\{main\}\);
> \# This dies

That’s because hash keys auto\-quote\.  You have to write
$table\{\+\_\_PACKAGE\_\_\}\, $table\{\(\_\_PACKAGE\_\_\)\} or something similar\.

The same applies\, e\.g\.\, to $table\{time\}\, which will use "time" rather
than "1372575833" as the hash key\.

That’s a feature\. :\-\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jkeenan

On 6/30/13 2&#8203;:53 AM\, Jim Avera wrote&#8203;:
> \# New Ticket Created by  Jim Avera
> \# Please include the string&#8203;:  \[perl \#118701\]
> \# in the subject line of all future correspondence about this issue\.
> \#\<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=118701>
>
>
>
> This is a bug report for perl from james\_avera@&#8203;yahoo\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.14\.2\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \#\!/usr/bin/perl
> use strict; use warnings;
>
> sub show\($\) \{ defined\($\_\[0\]\) ? \("'"\.$\_\[0\]\."'"\) : "undef" \}
>
> my %table = \( main =>  "the\-value" \);
>
> warn "\_\_PACKAGE\_\_ = "\,show\(\_\_PACKAGE\_\_\)\,"\\n";
> warn "table\{\_\_PACKAGE\_\_\} = "\, show\($table\{\_\_PACKAGE\_\_\}\)\, "\\n";
> warn "table\{'main'\} = "\, show\($table\{'main'\}\)\, "\\n";
>
> die "bug" unless show\($table\{\_\_PACKAGE\_\_\}\) eq show\($table\{main\}\);
> \# This dies
>

This limitation has been documented in 'perlmod' since 1997&#8203;:

\#\#\#\#\#
The special symbol C\<\_\_PACKAGE\_\_> contains the current package\, but 
cannot \(easily\) be used to construct variable names\.
\#\#\#\#\#

Is there any compelling reason why we should alter this long\-standing 
behavior?

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jimav

On 06/30/2013 05&#8203;:06 AM\, James E Keenan via RT wrote&#8203;:
> This limitation has been documented in 'perlmod' since 1997&#8203;: 
> \#\#\#\#\# The special symbol C\<\_\_PACKAGE\_\_> contains the current package\, 
> but cannot \(easily\) be used to construct variable names\. \#\#\#\#\# Is 
> there any compelling reason why we should alter this long\-standing 
> behavior?

At this point in Perl's history not much is really compelling\, so I 
guess the answer is no\.

But still\, perldata says \_\_PACKAGE\_\_ etc\. are "special literals" which 
may \(only\) be used as "separate tokens"\.  That makes it sound like they 
are recognized while scanning regardless of the context\.

I found this the hard way \(i\.e\. as a silent bug\) when $hash\{\_\_PACKAGE\_\_\} 
interpreted \_\_PACKAGE\_\_ as a bareword or unquoted string\, rather than as 
a special literal token\.  It's another "hidden gotcha" which would be 
nice to eliminate\.

\-Jim

**Comment by p5pRT** [comment]
### From @mauke

On 02\.07\.2013 00&#8203;:42\, Jim Avera wrote&#8203;:
> On 06/30/2013 05&#8203;:06 AM\, James E Keenan via RT wrote&#8203;:
>> This limitation has been documented in 'perlmod' since 1997&#8203;: \#\#\#\#\# The
>> special symbol C\<\_\_PACKAGE\_\_> contains the current package\, but cannot
>> \(easily\) be used to construct variable names\. \#\#\#\#\# Is there any
>> compelling reason why we should alter this long\-standing behavior?

Wrong part of the documentation\. The above only means you can't say 
$\_\_PACKAGE\_\_&#8203;::foo or similar\. Jim isn't trying to construct a variable name\.

> At this point in Perl's history not much is really compelling\, so I
> guess the answer is no\.
>
> But still\, perldata says \_\_PACKAGE\_\_ etc\. are "special literals" which
> may \(only\) be used as "separate tokens"\.  That makes it sound like they
> are recognized while scanning regardless of the context\.
>
> I found this the hard way \(i\.e\. as a silent bug\) when $hash\{\_\_PACKAGE\_\_\}
> interpreted \_\_PACKAGE\_\_ as a bareword or unquoted string\, rather than as
> a special literal token\.  It's another "hidden gotcha" which would be
> nice to eliminate\.

As Father Chrysostomos explained\, hash keys always autoquote 
identifiers\. Special\-casing \_\_PACKAGE\_\_ would introduce another hidden 
gotcha\, not eliminate it\.

\-\- 
Lukas Mai \<plokinom@&#8203;gmail\.com>

**Comment by p5pRT** [comment]
### From @b2gills

On Mon\, Jul 1\, 2013 at 10&#8203;:46 PM\, Lukas Mai \<plokinom@&#8203;gmail\.com> wrote&#8203;:
> On 02\.07\.2013 00&#8203;:42\, Jim Avera wrote&#8203;:
>>
>> On 06/30/2013 05&#8203;:06 AM\, James E Keenan via RT wrote&#8203;:
>>>
>>> This limitation has been documented in 'perlmod' since 1997&#8203;: \#\#\#\#\# The
>>> special symbol C\<\_\_PACKAGE\_\_> contains the current package\, but cannot
>>> \(easily\) be used to construct variable names\. \#\#\#\#\# Is there any
>>> compelling reason why we should alter this long\-standing behavior?
>
>
> Wrong part of the documentation\. The above only means you can't say
> $\_\_PACKAGE\_\_&#8203;::foo or similar\. Jim isn't trying to construct a variable name\.
>
>
>> At this point in Perl's history not much is really compelling\, so I
>> guess the answer is no\.
>>
>> But still\, perldata says \_\_PACKAGE\_\_ etc\. are "special literals" which
>> may \(only\) be used as "separate tokens"\.  That makes it sound like they
>> are recognized while scanning regardless of the context\.
>>
>> I found this the hard way \(i\.e\. as a silent bug\) when $hash\{\_\_PACKAGE\_\_\}
>> interpreted \_\_PACKAGE\_\_ as a bareword or unquoted string\, rather than as
>> a special literal token\.  It's another "hidden gotcha" which would be
>> nice to eliminate\.
>
>
> As Father Chrysostomos explained\, hash keys always autoquote identifiers\.
> Special\-casing \_\_PACKAGE\_\_ would introduce another hidden gotcha\, not
> eliminate it\.
>

If you need the special literal there are a few easy ways to do it

&nbsp;    perl \-E'%h=\(main=>"Hello World"\); say $h\{\+ \_\_PACKAGE\_\_ \}'
&nbsp;    perl \-E'%h=\(main=>"Hello World"\); say $h\{ \_\_PACKAGE\_\_\, \}'
&nbsp;    perl \-E'%h=\(main=>"Hello World"\); say $h\{ \( \_\_PACKAGE\_\_ \) \}'
&nbsp;    perl \-E'%h=\(main=>"Hello World"\); say $h\{ do\{ \_\_PACKAGE\_\_ \} \}'

basically anything that tells perl \(the compiler\) that you want
it treated as a statement will work\.

on 5\.16 or later you can also pretend it is a subroutine
&nbsp;    perl \-E'%h=\(main=>"Hello World"\); say $h\{ \_\_PACKAGE\_\_\(\) \}'

**Comment by p5pRT** [comment]
### From @jkeenan

On Mon Jul 01 22&#8203;:44&#8203;:37 2013\, brad wrote&#8203;:
> On Mon\, Jul 1\, 2013 at 10&#8203;:46 PM\, Lukas Mai \<plokinom@&#8203;gmail\.com> wrote&#8203;:
> > On 02\.07\.2013 00&#8203;:42\, Jim Avera wrote&#8203;:
> >>
> >> On 06/30/2013 05&#8203;:06 AM\, James E Keenan via RT wrote&#8203;:
> >>>
> >>> This limitation has been documented in 'perlmod' since 1997&#8203;: \#\#\#\#\#
> The
> >>> special symbol C\<\_\_PACKAGE\_\_> contains the current package\, but
> cannot
> >>> \(easily\) be used to construct variable names\. \#\#\#\#\# Is there any
> >>> compelling reason why we should alter this long\-standing behavior?
> >
> >
> > Wrong part of the documentation\. The above only means you can't say
> > $\_\_PACKAGE\_\_&#8203;::foo or similar\. Jim isn't trying to construct a
> variable name\.
> >
> >
> >> At this point in Perl's history not much is really compelling\, so I
> >> guess the answer is no\.
> >>
> >> But still\, perldata says \_\_PACKAGE\_\_ etc\. are "special literals"
> which
> >> may \(only\) be used as "separate tokens"\.  That makes it sound like
> they
> >> are recognized while scanning regardless of the context\.
> >>
> >> I found this the hard way \(i\.e\. as a silent bug\) when
> $hash\{\_\_PACKAGE\_\_\}
> >> interpreted \_\_PACKAGE\_\_ as a bareword or unquoted string\, rather
> than as
> >> a special literal token\.  It's another "hidden gotcha" which would
> be
> >> nice to eliminate\.
> >
> >
> > As Father Chrysostomos explained\, hash keys always autoquote
> identifiers\.
> > Special\-casing \_\_PACKAGE\_\_ would introduce another hidden gotcha\,
> not
> > eliminate it\.
> >
> 
> If you need the special literal there are a few easy ways to do it
> 
>     perl \-E'%h=\(main=>"Hello World"\); say $h\{\+ \_\_PACKAGE\_\_ \}'
>     perl \-E'%h=\(main=>"Hello World"\); say $h\{ \_\_PACKAGE\_\_\, \}'
>     perl \-E'%h=\(main=>"Hello World"\); say $h\{ \( \_\_PACKAGE\_\_ \) \}'
>     perl \-E'%h=\(main=>"Hello World"\); say $h\{ do\{ \_\_PACKAGE\_\_ \} \}'
> 
> basically anything that tells perl \(the compiler\) that you want
> it treated as a statement will work\.
> 
> on 5\.16 or later you can also pretend it is a subroutine
>     perl \-E'%h=\(main=>"Hello World"\); say $h\{ \_\_PACKAGE\_\_\(\) \}'
> 


So\, do we have a bug here?

A feature?

Something that needs better documentation?

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @b2gills

On Tue\, Jul 2\, 2013 at 5&#8203;:52 PM\, James E Keenan via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Mon Jul 01 22&#8203;:44&#8203;:37 2013\, brad wrote&#8203;:
>> On Mon\, Jul 1\, 2013 at 10&#8203;:46 PM\, Lukas Mai \<plokinom@&#8203;gmail\.com> wrote&#8203;:
>> > On 02\.07\.2013 00&#8203;:42\, Jim Avera wrote&#8203;:
>> >>
>> >> On 06/30/2013 05&#8203;:06 AM\, James E Keenan via RT wrote&#8203;:
>> >>>
>> >>> This limitation has been documented in 'perlmod' since 1997&#8203;: \#\#\#\#\#
>> The
>> >>> special symbol C\<\_\_PACKAGE\_\_> contains the current package\, but
>> cannot
>> >>> \(easily\) be used to construct variable names\. \#\#\#\#\# Is there any
>> >>> compelling reason why we should alter this long\-standing behavior?
>> >
>> >
>> > Wrong part of the documentation\. The above only means you can't say
>> > $\_\_PACKAGE\_\_&#8203;::foo or similar\. Jim isn't trying to construct a
>> variable name\.
>> >
>> >
>> >> At this point in Perl's history not much is really compelling\, so I
>> >> guess the answer is no\.
>> >>
>> >> But still\, perldata says \_\_PACKAGE\_\_ etc\. are "special literals"
>> which
>> >> may \(only\) be used as "separate tokens"\.  That makes it sound like
>> they
>> >> are recognized while scanning regardless of the context\.
>> >>
>> >> I found this the hard way \(i\.e\. as a silent bug\) when
>> $hash\{\_\_PACKAGE\_\_\}
>> >> interpreted \_\_PACKAGE\_\_ as a bareword or unquoted string\, rather
>> than as
>> >> a special literal token\.  It's another "hidden gotcha" which would
>> be
>> >> nice to eliminate\.
>> >
>> >
>> > As Father Chrysostomos explained\, hash keys always autoquote
>> identifiers\.
>> > Special\-casing \_\_PACKAGE\_\_ would introduce another hidden gotcha\,
>> not
>> > eliminate it\.
>> >
>>
>> If you need the special literal there are a few easy ways to do it
>>
>>     perl \-E'%h=\(main=>"Hello World"\); say $h\{\+ \_\_PACKAGE\_\_ \}'
>>     perl \-E'%h=\(main=>"Hello World"\); say $h\{ \_\_PACKAGE\_\_\, \}'
>>     perl \-E'%h=\(main=>"Hello World"\); say $h\{ \( \_\_PACKAGE\_\_ \) \}'
>>     perl \-E'%h=\(main=>"Hello World"\); say $h\{ do\{ \_\_PACKAGE\_\_ \} \}'
>>
>> basically anything that tells perl \(the compiler\) that you want
>> it treated as a statement will work\.
>>
>> on 5\.16 or later you can also pretend it is a subroutine
>>     perl \-E'%h=\(main=>"Hello World"\); say $h\{ \_\_PACKAGE\_\_\(\) \}'
>>
>
>
> So\, do we have a bug here?
>
> A feature?
>
> Something that needs better documentation?
>
> Thank you very much\.
> Jim Keenan
>
> \-\-\-
> via perlbug&#8203;:  queue&#8203;: perl5 status&#8203;: open
> https://rt-archive.perl.org/perl5/Ticket/Display.html?id=118701

This is an intended feature\, not a bug\.



---
As for documentation\, we always could use improvement

If we added a bit about how to use the special literal
\_\_PACKAGE\_\_ as a hash key\,
we would need to add it to every
keyword / op / subroutine that is in Perl\.

Maybe we could add it to perldata in the subscripts section\.

I don't think it's worth holding this ticket open for that though\.

 \-\-\-

There was a bit of discussion about allowing this to do
what beginners may think it does&#8203;:

&nbsp;    \_\_PACKAGE\_\_&#8203;::SubPackage

but that is not the topic of this bug report\.

**Comment by p5pRT** [comment]
### From @jimav

On 07/02/2013 03&#8203;:52 PM\, James E Keenan via RT wrote&#8203;:
>
> As Father Chrysostomos explained\, hash keys always autoquote
>> identifiers\.
>>> Special\-casing \_\_PACKAGE\_\_ would introduce another hidden gotcha\,
>> not
>>> eliminate it\.
>>>
>> If you need the special literal there are a few easy ways to do it
>>
>>      perl \-E'%h=\(main=>"Hello World"\); say $h\{\+ \_\_PACKAGE\_\_ \}'
>>      perl \-E'%h=\(main=>"Hello World"\); say $h\{ \_\_PACKAGE\_\_\, \}'
>>      perl \-E'%h=\(main=>"Hello World"\); say $h\{ \( \_\_PACKAGE\_\_ \) \}'
>>      perl \-E'%h=\(main=>"Hello World"\); say $h\{ do\{ \_\_PACKAGE\_\_ \} \}'
>>
>> basically anything that tells perl \(the compiler\) that you want
>> it treated as a statement will work\.
>>
>> on 5\.16 or later you can also pretend it is a subroutine
>>      perl \-E'%h=\(main=>"Hello World"\); say $h\{ \_\_PACKAGE\_\_\(\) \}'
>>
>> So\, do we have a bug here?
>>
>> A feature?
>>
>> Something that needs better documentation?
>> Thank you very much\.
>> Jim Keenan
IMO hidden or silent gotchas are the last significant 
software\-engineering risk specific to Perl\, now that the fundamental 
machinery works reliably \(and btw\, thanks immensely to all who made that 
happen\!\)

The special \_\_\(PACKAGE|FILE|LINE\)\_\_ symbols are well\-documented and 
vanishingly unlikely to be innocently used as an application symbol\.  So 
it's hard to see how special\-casing them to always be treated as string 
literals with the implied values \(and never auto\-quote themselves\) would 
make things worse\.    Removing hidden or silent traps is always a good 
thing in the abstract\.

But I'm not familiar with perl internals\, so doing so might be 
excessively risky\.

I think those who would do the work are the only ones qualified to say 
whether fixing this is prudent\.

**Comment by p5pRT** [comment]
### From @ikegami

On Tue\, Jul 2\, 2013 at 9&#8203;:09 PM\, Jim Avera \<james\_avera@&#8203;yahoo\.com> wrote&#8203;:

> The special \_\_\(PACKAGE|FILE|LINE\)\_\_ symbols are well\-documented and
> vanishingly unlikely to be innocently used as an application symbol\.  So
> it's hard to see how special\-casing them to always be treated as string
> literals with the implied values \(and never auto\-quote themselves\) would
> make things worse\.


You could say exactly the same for \`split\`\, \`uc\`\, etc

And that's why it would make things worse\.

**Comment by p5pRT** [comment]
### From @rjbs


We're not changing this\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]


We're not changing this\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
@rjbs - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From @jimav

On 07/03/2013 04&#8203;:08 AM\, Eric Brine wrote&#8203;:
> On Tue\, Jul 2\, 2013 at 9&#8203;:09 PM\, Jim Avera \<james\_avera@&#8203;yahoo\.com 
> \<mailto&#8203;:james\_avera@&#8203;yahoo\.com>> wrote&#8203;:
>
>     The special \_\_\(PACKAGE|FILE|LINE\)\_\_ symbols \.\.\. it's hard to see
>     how special\-casing them to always be treated as string literals
>     with the implied values \(and never auto\-quote themselves\) would
>     make things worse\. 
>
>
> You could say exactly the same for \`split\`\, \`uc\`\, etc

Ok\, I see your point\.   Death\, taxes\, and ambiguous programming languages\.
